core/src/main/java/com/datastax/oss/driver/internal/core/loadbalancing/RackAwareWeightedLoadBalancingPolicy.java

```diff
@@ -0,0 +1,264 @@
+ */
+public class RackAwareWeightedLoadBalancingPolicy extends DefaultLoadBalancingPolicy {
+  private static final Logger LOG =
+      LoggerFactory.getLogger(RackAwareWeightedLoadBalancingPolicy.class);
+  // Each client will randomly skew so traffic is introduced gradually to a newly up replica
+  // Each client will start sending at a period between 15s and 30, and they will gradually
+  // increase load over the next 15 seconds.
+  private static final long DELAY_TRAFFIC_SKEW_MILLIS = SECONDS.toMillis(15);
+  private static final long DELAY_TRAFFIC_MILLIS =
+      DELAY_TRAFFIC_SKEW_MILLIS + ThreadLocalRandom.current().nextLong(DELAY_TRAFFIC_SKEW_MILLIS);
+
+  // By default we will only score this many nodes, the rest will get added on without scoring.
+  // We don't usually need to score every single node if there are more than a few.
+  static final int DEFAULT_SCORED_PLAN_SIZE = 8;
+  // Default multiplicative weights. Think of this like "How much concurrency must occur
+  // before I fail off this node to the next". Note that these defaults are intentionally
+  // meant to shed load to unloaded rack coordinators when a replica set is all
+  // relatively heavily loaded (specifically 3x as loaded).
+  static final double DEFAULT_WEIGHT_NON_RACK = 4.0;
+  static final double DEFAULT_WEIGHT_NON_REPLICA = 12.0;
+  static final double DEFAULT_WEIGHT_STARTING = 16.0;
+  static final double DEFAULT_WEIGHT_UNHEALTHY = 64.0;
+
+  private final int planSize;
+  private final double weightNonRack;
+  private final double weightNonReplica;
+  private final double weightStarting;
+  private final double weightUnhealthy;
+
+  public RackAwareWeightedLoadBalancingPolicy(
+      @NonNull DriverContext context,
+      @NonNull String profileName) {
+    super(context, profileName);
+    this.planSize = profile.getInt(DefaultDriverOption.LOAD_BALANCING_SCORED_PLAN_SIZE, DEFAULT_SCORED_PLAN_SIZE);
+    // Choices of weights will change how this load balancer prefers endpoints.
+    // The weight is relative to the outstanding concurrency.
+    this.weightNonRack = profile.getDouble(
+        DefaultDriverOption.LOAD_BALANCING_WEIGHT_NON_RACK, DEFAULT_WEIGHT_NON_RACK);
+    this.weightNonReplica = profile.getDouble(
+        DefaultDriverOption.LOAD_BALANCING_WEIGHT_NON_REPLICA, DEFAULT_WEIGHT_NON_REPLICA);
+    this.weightStarting = profile.getDouble(
+        DefaultDriverOption.LOAD_BALANCING_WEIGHT_STARTING, DEFAULT_WEIGHT_STARTING);
+    this.weightUnhealthy = profile.getDouble(
+        DefaultDriverOption.LOAD_BALANCING_WEIGHT_UNHEALTHY, DEFAULT_WEIGHT_UNHEALTHY);
+  }
+
+  @NonNull
+  @Override
+  public Queue<Node> newQueryPlan(Request request, Session session) {
+    if (session == null) {
+      return super.newQueryPlan(request, null);
+    }
+
+    // Take a copy of nodes and reference to replicas since the node map is concurrent
+    Set<Node> dcNodeSet = getLiveNodes().dc(getLocalDatacenter());
```

Comment:
If a user sets a local DC, the current implementation only plans and scores for Cassandra hosts in the local DC. If a user doesn't set a local DC, there's no weighting to encourage the query plan to select a coordinator in the local DC over one in a remote DC, right? If there's no local DC configured, `getLocalDatacenter` is empty, and `dcNodeSet` ends up being all nodes.

If a user uses all non-local consistency levels (QUORUM, etc.) and doesn't want all traffic to go to a coordinator in the local DC but does want to encourage the query planner to pick coordinators in the local DC, how would they configure that?

Reply:
Currently the RackAwareWeightedLoadBalancingPolicy extends DefaultLoadBalancingPolicy, which uses `MandatoryLocalDcHelper` and so requires a local DC to be set. I can however see an argument where it instead extends `BasicLoadBalancingPolicy`, in a world where RAWLBP is a more general solution with sensible defaults (kinda like UCS is for Cassandra compaction?) and you can tweak weights to get what you want.

Regarding non-local consistency level - if we allow scoring non-local-DC nodes, I think we'd still want to weight the local ones higher and rely on the server-side snitch to figure out a quorum partner if needed.